### PR TITLE
Removed all imports from barrel files for components, actions, helpers, hooks and lodash

### LINKED ISSUE #6507: Remove any usage of barrel files from core
## PLIP (Plone Improvement Proposal)

This is a first step for #5419 and for deprecating the usage of barrel files entirely. Eventually, those files should be removed in Plone 7.

## Abstract

Barrel files create issues in code splitting and bundling, especially because we currently use an older bundler, but even modern bundlers have a hard time with those. [Several articles and examples can be found online](https://www.google.com/search?q=removing+barrel+files)

## Motivation

Solving this issue is mandatory to improve performance and eventually it is already decided that those files should not exist in Plone 7.

## Assumptions

Volto 18 should have the files for backwards compatibility, the usage of those should just be avoided.

## Proposal & Implementation

Update every import in core volto. Add a eslint rule that raises a warning for each import from barrel files. Update the docs to reflect this choice and guide the migration.

## Deliverables

for Plone 6.1 / Volto 18:
- [x] Update Volto to avoid using barrel files internally (#6509)
- [x] Add info about this deprecation in the upgrade guide (#6426)

for Plone 7:
- [ ] Fixer to help remove barrel imports from add-ons?

Initial work on a codemod for automating the migration:
https://github.com/plone/volto/blob/7b047fb4b2d71395ddba05b7e2570fcec70fc8d4/packages/volto/src/replace-barrel-imports.cjs

- [ ] Remove the unused barrel files.

## Risks

No risks involved. This has to be fully backwards compatible.

## Participants

@pnicolli 
@sneridagh 
@stevepiercy

### PR DESCRIPTION
**Pull request by pnicolli**
Closes #6507

### CODE REVIEW

File: packages/volto/.eslintrc [hunk 1]
```diff
@@ -24,33 +25,50 @@
...
+      },
+      {
+        "name": "@plone/volto/helpers",
+        "message": "Importing from barrel files is not allowed. Please use direct imports of the modules instead."
+      },
+      {
+        "name": "@plone/volto/actions",
+        "message": "Importing from barrel files is not allowed. Please use direct imports of the modules instead."
+      }
+    ],
```
Comment (pnicolli) [inline packages/volto/.eslintrc:52]: @sneridagh this eslint setup for some reason still doesn't work in the editor for me so I could properly try this.

**Comment by pnicolli** [comment]
@stevepiercy I realized the deprecation docs for this already existed, I added a couple extra files I cleaned up.

**Comment by pnicolli** [comment]
FYI: this PR also reduces the main `client` bundle size by almost 30kb gzipped! Just by removing the imports from barrel files!

File: packages/volto-slate/news/6509.internal [hunk 2]
```diff
@@ -0,0 +1 @@
+Removed all imports from barrel files for components, actions, helpers, hooks and lodash @pnicolli
```
Comment (stevepiercy) [inline packages/volto-slate/news/6509.internal:1]:
```suggestion
Removed all imports from barrel files for components, actions, helpers, hooks and lodash. @pnicolli
```

File: packages/volto/news/6509.internal [hunk 3]
```diff
@@ -0,0 +1 @@
+Removed all imports from barrel files for components, actions, helpers, hooks and lodash @pnicolli
```
Comment (stevepiercy) [inline packages/volto/news/6509.internal:1]:
```suggestion
Removed all imports from barrel files for components, actions, helpers, hooks and lodash. @pnicolli
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Looks good, just add a couple of periods, then I approve docs and news.

Out of curiosity, did you do all this manually or with a tool that finds barrel imports and replaces them? This is a huge amount of changes!

**Comment by pnicolli** [comment]
> Out of curiosity, did you do all this manually or with a tool that finds barrel imports and replaces them? This is a huge amount of changes!

@stevepiercy All manually :) I figured that I would have had a lot of 30 to 60 minutes slots during the conf sprint, the following day, the flights and the long layovers, so this task would be a good fit

**Review by stevepiercy — APPROVED** [review]
Docs and news approved. Still needs tech review.

File: packages/volto/.eslintrc (hunk 1 above)
Comment (sneridagh) [reply packages/volto/.eslintrc]: @pnicolli they were in the wrong place! Now they are fine.

**Comment by stevepiercy** [comment]
Needs an ESLint fix, then should be good to merge.

**Comment by sneridagh** [comment]
@pnicolli now should be green, let's merge it and release it.
Thanks for taking care, it rocks!

**Comment by sneridagh** [comment]
@pnicolli just realised of something: 
if we leave the ESlint rules "on" by default, they will break all projects outside, so we cannot enable it by default in 18.
:/ MEH!

**Review by sneridagh — APPROVED** [review]
LGTM and ready to merge. 

@pnicolli waiting for your last ack.

**Comment by davisagli** [comment]
@sneridagh @pnicolli Are we merging this for Volto 18 or 19? If it's for 18, it's a bit late (we already made non-alpha releases and this is breaking). If it's for 19, fine, but then we need to cut an 18.x branch and start using main for 19.

**Comment by davisagli** [comment]
@sneridagh @pnicolli Oh sorry, now that I actually read the PLIP, I understand that this is fixing the imports in Volto core but not actually removing the barrels yet. Okay!

**Comment by pnicolli** [comment]
I found another barrel file that I didn't see in the first iteration, I worked on it last night, will finish asap.
It is `@plone/volto-slate/editor/ui`

**Comment by sneridagh** [comment]
@davisagli yes, we realised that the work of moving all the imports to direct imports is fine and non-breaking. This will also enable right now multi-framework add-ons :)

**Comment by pnicolli** [comment]
Ok added the last small commit, now I think it's good to go @sneridagh

**Pull request merged**